### DISCUSSION TITLE
Docker finn build 관련 질문

### DISCUSSION THREAD

**Discussion by WooYoung1999** [pbk20191/embedded-social-infra#4] [Q&A] [answered]
@pbk20191 
안녕하세요. 지난번에 보여주신 docker finn build 중 오류가 발생했는데 혹시 알고 계신지 하여 여쭙니다.
<img width="441" alt="docker 오류" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/f5d17de8-6d0d-4345-9f69-22f43ea32678">
<img width="719" alt="docker 오류2" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/79573b3f-7aa7-40b2-838f-a4dfedc511c3">

wsl 내부에서 pytorch 패키지를 다운로드하는 중에 오류가 생긴 것으로 보이는데 exit code:137이 docker 내부 메모리가 부족하여 생기는 오류(Out Of Memory)라고 뜹니다. 그런데 docker 컨테이너를 확인해보니 현재 docker에 존재하는 컨테이너가 없어서 메모리를 확인할 수가 없었습니다. 제가 생각하기에 pytorch패키지의 용량이 커서 다운로드하다 멈춘 것으로 보이는데 혹시 해결법을 아시는지 여쭤봅니다.
참고로 wsl내부에서 쓰고 있는 python3의 버전은 3.10.12입니다.
<img width="352" alt="docker 오류3" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/f1825991-105e-43dc-844c-96c6611ee267">

**Comment by pbk20191** [comment]
@WooYoung1999 
2가지 오류가 보이네요.
- XRT 런타임 설치 실패
- 파이토치 설치실패

1. 음... 제 기억으로는 노트북이 CUDA 지원하지 않는 노트북으로 알고 있는데, CPU 토치가 아니라 CUDA 토치를 다운로드하려고 하네요.
노트북이 CUDA를 지원하는 노트북인가요? 그리고 저 빨간색으로 밑줄 친 로그가 나오는지 확인해주세요. CUDA가 없는데 저 로그가 나오면 안되긴 합니다.
<img width="494" alt="스크린샷 2024-04-14 오후 1 53 08" src="https://github.com/pbk20191/embedded-social-infra/assets/29122714/23c24b60-f3d1-4a26-be00-0c4809b95c71">

2. 전체 디스크 용량 확보
일단 도커 이미지만 해도 약 9.31 ~ 10.46GB 정도 나오고, 여기에 도커 빌드 캐시 생각하면 최소 20GB이상 여유 공간은 있어야 할 것이라 추측이 됩니다. ( 제 의견으로 권장은 30GB 이상입니다.)
거기에 finn이 자체도 이것저것하면 용량이 1.1 ~ 1.2GB 정도 나옵니다.
따라서 여기에 대응하는 용량 확보를 하시는것이 좋을 것 같습니다. 만약 용량 확보가 어렵다면 적당한 외장하드나 SSD를 구해서, Docker Desktop에 Mount해야 합니다. Docker Desktop -> settings -> Resources -> Disk Image Location
<img width="565" alt="image" src="https://github.com/pbk20191/embedded-social-infra/assets/29122714/bc2e97c8-3a23-40d0-bfa0-64a8b1ac1262">

3. XRT 런타임은 Edge Device target으로 할 경우에 사용되지는 않으나, 실패자체가 도커 캐싱에 방해하므로 일단 고칠 필요가 있네요.

도커 로그상에서 "https://www.xilinx.com/bin/public/openDownload?filename=" 이런 URL에 대한 다운로드가 성공했는지 실패했는지 확인하시고, 만약에 다운로드가 실패한거면 timezone문제 일 확률이 있습니다. 이것은 finn 폴더에서 docker/Dockerfile.finn을 수정해야 합니다.
확인해보시고 아래 이미지처럼 타임존을 수정해야 할 수도 있어요.
<img width="602" alt="스크린샷 2024-04-14 오후 2 18 35" src="https://github.com/pbk20191/embedded-social-infra/assets/29122714/09abd4e4-a567-483b-a9d1-522d0dad5019">
XRT런타임 다운로드 링크는 아래와 같을 텐데, 접속해서 다운로드 되는지도 확인해보시고요.
https://www.xilinx.com/bin/public/openDownload?filename=xrt_202220.2.14.354_22.04-amd64-xrt.deb

4. 자세한 정보를 얻을 수 있도록 아래의 사진과 같이 Docker Desktop 에서 Build Tab을 확인하면 Failed Build를 볼 수 있을건데 이것을 눌러서 이것에 대한 로그를 첨부해주세요.
<img width="558" alt="스크린샷 2024-04-14 오후 1 44 41" src="https://github.com/pbk20191/embedded-social-infra/assets/29122714/8ccd9fef-6a2b-451f-b104-d1ed99d6b61f">

**Reply by WooYoung1999** [reply]
@pbk20191
1. 제 노트북은 CUDA를 지원하지 않고 로그 상에서 밑줄쳐진 부분은 없습니다.
<img width="369" alt="vscode log" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/e73ccb5b-db4e-4499-ab6d-fa7148f418ba">

2. 제 C드라이브 용량이 약 13GB밖에 남지 않아서 지금 디스크 용량 정리중입니다. 허나 만약에 C드라이브를 정리해도 용량이 충분치 않으면 Docker 내부에서 이미지 저장 경로를 D드라이브로 옮겨야 할 것 같은데 다음과 같이 경로만 지정해도 괜찮을까요? 구글링해보니 추가적으로 도커 바인드 마운트를 설정하는 과정이 있긴합니다.
docker container run --mount type=bind,source=$source,target=$target
<img width="365" alt="path" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/e7cb56a0-72a3-4113-bdf9-2dfa0ab68c91">

3. 말씀하신 URL다운로드는 정상적으로 수행된 것 같습니다. timezone문제는 아니라고 생각합니다.
<img width="543" alt="docker log2" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/a9245d00-3ce8-49d8-adbb-c85661748306">

4. 요청해주신 failed build의 LOG를 첨부합니다.
<img width="1064" alt="docker log2" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/b247942c-0c18-40a5-b350-d9d4cd52f12d">

**Reply by pbk20191** [reply]
2. `docker container run --mount type=bind,source=$source,target=$target` 이 커맨드는 도커 컨테이너에다가 호스트의 디렉토리를 링크할때 쓰는 것이지 우리가 사용하는 용도에는 맞지 않습니다. 따라서 보여주신 것처럼 D드라이브로 도커 이미지 경로를 변경하고 도커를 재시작하면 될 것 같습니다.

다시확인해보니 xrt는 정상적으로 설치가 된 것 같네요.

=> 도커 이미지 경로를 D 드라이브 경로로 바꾸고, 도커 재시작한 다음에 다시 진행하면 될 것 같습니다.

**Reply by WooYoung1999** [reply]
여전히 pytorch를 다운로드하는 과정에서 같은 오류가 발생하네요... 혹여나 RAM의 부족으로 프로그램이 강제종료되는 것인가 하여 램 용량을 일정 비우고 다시 시도해보았는데도 같은 오류가 발생합니다. 전체 코드 중에 pytorch 패키지를 다운로드 하는 과정이 포함되어 있는데 그렇다면 docker hub에서 pytorch 패키지를 미리 설치해놓고 해당 코드를 돌린다면 오류가 발생하지 않을까요?

**Reply by pbk20191** [reply, marked as answer]
어처피 cuda사용 못하기도 하고, 3GB 바이너리 다운로드하다가 터지기도 하니까
pytorch line에서  `--extra-index-url https://download.pytorch.org/whl/cu116`를 `--extra-index-url https://download.pytorch.org/whl/cpu`로 바꿔보실래요?

**Reply by pbk20191** [reply]
아 finn/docker/Dockerfile.finn 파일의 line 92번에 있습니다.

**Reply by WooYoung1999** [reply]
된것 같습니다. cpu로 바꾸니까 용량이 1/10로 줄어버리네요... 
<img width="1049" alt="build성공1" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/6791a7fd-55b9-4417-b783-8939ce3a3b90">
<img width="932" alt="build성공2" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/e8c053a4-8ff3-4f89-932e-5963a83baa2e">
<img width="1005" alt="build성공3" src="https://github.com/pbk20191/embedded-social-infra/assets/156506148/2a6471f8-7eb9-4ef1-93b4-607bfa95f5bc">
